src/couch/src/couch_db_updater.erl

```diff
@@ -857,17 +864,23 @@ merge_rev_tree(OldInfo, NewDoc, Client, Limit, false) ->
             send_result(Client, NewDoc, conflict),
             OldInfo
     end;
-merge_rev_tree(OldInfo, NewDoc, _Client, Limit, true) ->
+merge_rev_tree(OldInfo, NewDoc, _Client, true) ->
     % We're merging in revisions without caring about
     % conflicts. Most likely this is a replication update.
     OldTree = OldInfo#full_doc_info.rev_tree,
     NewTree0 = couch_doc:to_path(NewDoc),
-    {NewTree, _} = couch_key_tree:merge(OldTree, NewTree0, Limit),
+    {NewTree, _} = couch_key_tree:merge(OldTree, NewTree0),
     OldInfo#full_doc_info{rev_tree = NewTree}.
 
-stem_full_doc_infos(#db{revs_limit=Limit}, DocInfos) ->
-    [Info#full_doc_info{rev_tree=couch_key_tree:stem(Tree, Limit)} ||
-            #full_doc_info{rev_tree=Tree}=Info <- DocInfos].
+stem_full_doc_info(#full_doc_info{rev_tree = Tree} = Info, Limit) ->
+    Stemmed = couch_key_tree:stem(Tree, Limit),
+    Info#full_doc_info{rev_tree = Stemmed}.
+
+full_stem_full_doc_infos(#db{revs_limit=Limit}, DocInfos) ->
```

Comment:
Because I still have nightmares of COUCHDB-1163 and didn't want to throw away our function that automatically cleans up broken revision trees due to bad merges. I'm hand waving a bit that its run in the compactor only now and hopefully it doesn't cause us too much headache (though its a huge possibility that interactive conflict generation now becomes fast enough that we generate trees that are so bad that compaction can get past them).

I'm also mulling over slightly different approaches to stemming that use dicts or some other such approach but that gets us into O(N log N) territory.

Reply:
To clarify that last bit, I'm contemplating other levels of stemming options that may be slightly slower than the current approach while still getting us better guarantees on correct trees due to merging issues. I've also contemplated different approaches to merging that would slow that operation down slightly while hopefully being more correct and not relying on a correct order of application.